### PR TITLE
[youtube] Add options to skip webpage and js player for video extraction

### PR DESCRIPTION
**Pull request by coletdjnz**
## Please follow the guide below

- You will be asked some questions, please read them **carefully** and answer honestly
- Put an `x` into all the boxes [ ] relevant to your *pull request* (like that [x])
- Use *Preview* tab to see how your *pull request* will actually look like

---

### Before submitting a *pull request* make sure you have:
- [X] At least skimmed through [adding new extractor tutorial](https://github.com/ytdl-org/youtube-dl#adding-support-for-a-new-site) and [youtube-dl coding conventions](https://github.com/ytdl-org/youtube-dl#youtube-dl-coding-conventions) sections
- [X] [Searched](https://github.com/yt-dlp/yt-dlp/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [X] Checked the code with [flake8](https://pypi.python.org/pypi/flake8)

### In order to be accepted and merged into yt-dlp each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check one of the following options:
- [X] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [X] (Future?) Bug fix
- [X] Improvement
- [ ] New extractor
- [ ] New feature

---

### Description of your *pull request* and other information

Not sure what to call this.

This code was originally from  #682.

This PR adds:
- fallback player url extraction when it fails to be extracted from the webpage
- don't download js player unnecessarily for clients that don't require it
- try extract js player url from any additional client configs
- ability to skip the js player usage/download (inc iframe fallback).
    -  This overrides the clients' config setting.
    -  **This can be enabled with `--extractor-args youtube:player_skip=js`**
    -  Formats that require signature decryption will be skipped automatically (since no player url)
- ability to skip the webpage download
    -  **This can be enabled with `--extractor-args youtube:player_skip=webpage`**

KNOWN ISSUES:
- authentication for multi-channel accounts and multi-account cookies may not work correctly if the webpage (and any additional client configs) are not downloaded. 

TODO:
- documentation
- `player_skip=js` or something else?

### CODE REVIEW

**Comment by coletdjnz** [comment]
Some example usage

1. `--extractor-args youtube:player_skip=webpage,configs,playerjs;player_client=android,web`
- This makes extraction solely rely on the API - so no webpage downloads. Might help with avoiding 429s in some situations.
- Since the web client usually needs the player js, formats from the web client requiring signature decryption will be skipped.
- 3 initial API requests (android, web, initial data)

2. `--extractor-args youtube:player_skip=webpage,configs,playerjs;player_client=android`
- Same as above, but without the web client (some minor metadata may be missed - be sure to check)
- 2 initial API requests (android, initial data)

3. `--extractor-args youtube:player_skip=webpage`
- basically forces fallbacks as if the webpage failed to download.
- player url should be downloaded using the iframe API fallback (once, only if required)

4. `--extractor-args youtube:player_skip=webpage;player_client=web_music`
- config webpage for web_music is downloaded (thus player js details), so web formats should work correctly.

Not recommended :
1. `--extractor-args youtube:player_skip=webpage,configs,playerjs;player_client=web`
- In some cases you'll get a no formats error (formats from the web client requiring signature decryption will be skipped).
2. `--extractor-args youtube:player_skip=playerjs;player_client=web`

File: yt_dlp/extractor/youtube.py [hunk 1]
```diff
@@ -2702,17 +2737,19 @@ def _real_extract(self, url):
...
         master_ytcfg = self.extract_ytcfg(video_id, webpage) or self._get_default_ytcfg()
-        player_url = self._extract_player_url(master_ytcfg, webpage)
         identity_token = self._extract_identity_token(webpage, video_id)
 
         player_responses = list(self._extract_player_responses(
             self._get_requested_clients(url, smuggled_data),
-            video_id, webpage, master_ytcfg, player_url, identity_token))
+            video_id, webpage, master_ytcfg, identity_token))
 
+        player_url = player_responses.pop(-1)
```
Comment (pukkandan) [inline yt_dlp/extractor/youtube.py:2752]:
```suggestion
        player_url = player_responses.pop()
```
why -1?

Comment (coletdjnz) [reply yt_dlp/extractor/youtube.py:2752]: :sweat_smile:

File: yt_dlp/extractor/youtube.py [hunk 2]
```diff
@@ -2452,11 +2469,11 @@ def _is_agegated(player_response):
     def _is_unplayable(player_response):
         return traverse_obj(player_response, ('playabilityStatus', 'status')) == 'UNPLAYABLE'
 
-    def _extract_player_response(self, client, video_id, master_ytcfg, player_ytcfg, identity_token, player_url, initial_pr):
+    def _extract_player_response(self, client, video_id, master_ytcfg, player_ytcfg, identity_token, player_url, initial_pr, require_js_player):
 
         session_index = self._extract_session_index(player_ytcfg, master_ytcfg)
         syncid = self._extract_account_syncid(player_ytcfg, master_ytcfg, initial_pr)
-        sts = self._extract_signature_timestamp(video_id, player_url, master_ytcfg, fatal=False)
+        sts = self._extract_signature_timestamp(video_id, player_url, master_ytcfg, fatal=False) if require_js_player else None
```
Comment (pukkandan) [inline yt_dlp/extractor/youtube.py:2476]: Is there really a need for `require_js_player`? Can't we just check if `player_url is None`? I also noticed `fatal=true` is never used for `_extract_signature_timestamp`

File: yt_dlp/extractor/youtube.py [hunk 3]
```diff
@@ -2527,15 +2544,27 @@ def append_client(client_name):
...
 
+            player_url = player_url or self._extract_player_url(master_ytcfg, player_ytcfg, webpage=webpage)
+            require_js_player = self._get_default_ytcfg(client).get('REQUIRE_JS_PLAYER')
+            if 'js' in self._configuration_arg('player_skip'):
+                require_js_player = False
+                player_url = None
+
+            if not player_url and not tried_iframe_fallback and require_js_player:
+                player_url = self._download_player_url(video_id)
+                tried_iframe_fallback = True
```
Comment (pukkandan) [inline yt_dlp/extractor/youtube.py:2563]: assuming `require_js_player` doesnt need to be passed to `_extract_player_response`, this can be simplified

File: yt_dlp/extractor/youtube.py [hunk 4]
```diff
@@ -2552,6 +2581,7 @@ def append_client(client_name):
             elif self._is_agegated(pr):
                 append_client(f'{client}_agegate')
 
+        yield player_url
```
Comment (pukkandan) [inline yt_dlp/extractor/youtube.py:2584]: yielding a totally unrelated value is confusing. I believe it is better to simply forego the laziness here. (It was made lazy only for code simplicity and `list` is being called on the output immediately)

File: yt_dlp/extractor/youtube.py (hunk 2 above)
Comment (coletdjnz) [reply yt_dlp/extractor/youtube.py:2476]: require js player: for the case where we have a player url but the particular client does not require it, from memory. Though thinking now we could remove that parameter and just do `player_url=player_url if require_js_player else None` on the func call.

I think the fatal thing was just to keep consistency with the signature extraction function originally. Could probably be set to be not fatal by default imo.

Comment (pukkandan) [reply yt_dlp/extractor/youtube.py:2476]: > require js player: for the case where we have a player url but the particular client does not require it, from memory. Though thinking now we could remove that parameter and just do player_url=player_url if require_js_player else None on the func call.

Unless I am mistaken, the below code ensures that `player_url` is never set if not `require_js_player` 

> I think the fatal thing was just to keep consistency with the signature extraction function originally. Could probably be set to be not fatal by default imo

I meant that the function is never called with `fatal=True` anywhere in the code. So I was thinking that whole parameter can be scraped. But if you think it will be usefull in future, let it be

Comment (coletdjnz) [reply yt_dlp/extractor/youtube.py:2476]: > Unless I am mistaken, the below code ensures that player_url is never set if not require_js_player

nah. `player_url` is only cleared when the player_skip arg is given. We don't clear it otherwise since we might have other clients in the iteration that may need it. `require_js_player` just tells us if we need to worry about having a player (and sts) for the current client in the iteration.

File: README.md [hunk 5]
```diff
@@ -1436,7 +1436,10 @@ The following extractors use this feature:
 * **youtube**
     * `skip`: `hls` or `dash` (or both) to skip download of the respective manifests
     * `player_client`: Clients to extract video data from. The main clients are `web`, `android`, `ios`, `mweb`. These also have `_music`, `_embedded`, `_agegate`, and `_creator` variants (Eg: `web_embedded`) (`mweb` has only `_agegate`). By default, `android,web` is used, but the agegate and creator variants are added as required for age-gated videos. Similarly the music variants are added for `music.youtube.com` urls. You can also use `all` to use all the clients
-    * `player_skip`: `configs` - skip any requests for client configs and use defaults
+    * `player_skip`:
+        - `configs` - skip any requests for client configs and use defaults
+        - `webpage` - skip request for the initial webpage
+        - `js` - skip request & usage of the js player file for all clients
```
Comment (pukkandan) [inline README.md:1442]:
```suggestion
    * `player_skip`: One or more of `configs` (skip requests for client configs), `webpage` (skip request for the initial webpage), `js` (skip request for js player). While these options can help reduce the number of requests needed or avoid some rate-limiting, they could cause some issues. See [#860](https://github.com/yt-dlp/yt-dlp/pull/860) for more info
```
